### PR TITLE
register default functions for federation/apis/core

### PR DESCRIPTION
**Pull request by mfanjie**
fix issue #35667 and all related e2e test fail on federation.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35765)

### CODE REVIEW

**Comment by mfanjie** [comment]
@quinton-hoole @nikhiljindal FYI. all e2e failure related to missing optional properties should be fixed.

**Comment by nikhiljindal** [comment]
FWIW, our federated service test has continuously been passing: https://k8s-testgrid.appspot.com/google-federation. Can you point out which specific e2e test are you fixing?

cc @kubernetes/sig-cluster-federation

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

**Comment by mfanjie** [comment]
this pr fixed the federation API incompatible issue that @quinton-hoole suggested me to follow, which service can not be created as the default function for federation/apis/core/v1 objects were not registered. Check the detail for the linked issue.

???? iPhone

? 2016?10?29??03:14?k8s-ci-robot <notifications@github.com<mailto:notifications@github.com>> ???


Jenkins GCE Node e2e failed<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/35765/pull-kubernetes-node-e2e/2131/> for commit c2cec01<https://github.com/kubernetes/kubernetes/commit/c2cec013de4396741c7ee4840755b0b50cce18d0>. Full PR test history<http://pr-test.k8s.io/35765>.

The magic incantation to run this job again is @k8s-bot node e2e test this. Please help us cut down flakes by linking to an open flake issue<https://github.com/kubernetes/kubernetes/issues?q=is:issue+label:kind/flake+is:open> when you hit one in your PR.

-

**Comment by mfanjie** [comment]
@nikhiljindal are you saying that the API incompatible issues were resolved by other PR?
This PR fixed the issue you posted on #28615
The error of missing optional properties is because of default functions are not registered.
```
Unexpected error occurred: Service "federated-service" is invalid: [spec.ports[0].protocol: Required value, spec.sessionAffinity: Required value, spec.ports[0].protocol: Invalid value: "": cannot create an external load balancer with non-TCP/UDP ports]
```

**Comment by mfanjie** [comment]
Jenkins GCE Node e2e failed, and it seems a flake 
`I1028 12:12:56.128] E1028 19:12:54.860058    1610 services.go:279] failed to get "cloud-init.log" from journald: Failed to add filter for units: No data available
I1028 12:12:56.128] , exit status 1`

**Comment by nikhiljindal** [comment]
Yes, the test was failing due to some other issue as it is green now: https://k8s-testgrid.appspot.com/google-federation.

The change seems fine to me as per https://github.com/kubernetes/kubernetes/pull/33352.
But if this change is really needed, then we should also add a test which fails without this change.

cc @smarterclayton as author of https://github.com/kubernetes/kubernetes/pull/33352.

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

File: federation/apis/core/v1/defaults.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ import (
 )
 
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	v1.RegisterDefaults(scheme)
```
Comment (smarterclayton) [inline federation/apis/core/v1/defaults.go]: Yes, if you include upstream objects you must also include their defaulters.  Once you do this though you no longer need the next line.

**Comment by mfanjie** [comment]
@nikhiljindal  I did not direct link of the error to this PR, can you please re run the test?

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

**Comment by nikhiljindal** [comment]
LGTM

**Comment by nikhiljindal** [comment]
@k8s-bot node e2e test this

**Comment by saad-ali** [comment]
Release-czar approved post-code freeze merge--This was LGTMed and in the merge-queue at code freeze time for 1.5. Adding 1.5 milestone to let it gets merged after code freeze.

**Pull request merged**